File: inventory/internal/infra/messaging/relay.go
package messaging

import (
	"context"
	"errors"
	"fmt"
	"time"

	"inventory/internal/model"

	amqp "github.com/rabbitmq/amqp091-go"
)

const (
	relayInterval  = time.Second
	relayBatch     = 50
	relayLease     = 30 * time.Second
	publishTimeout = 5 * time.Second

	backoffBase = 2 * time.Second
	backoffCap  = time.Minute
)

var (
	errBrokerRefused = errors.New("broker refused the message")
	errUnroutable    = errors.New("no queue bound to the routing key")
)

type Relay struct {
	repository model.OutboxRepository
	exchange   string
	connection *Connection
}

func NewRelay(repository model.OutboxRepository, exchange string) *Relay {
	return &Relay{
		repository: repository,
		exchange:   exchange,
	}
}

func (r *Relay) Start() {
	go func() {
		ticker := time.NewTicker(relayInterval)
		defer ticker.Stop()

		for range ticker.C {
			r.drain()
		}
	}()

	fmt.Println("Outbox relay started, publishing to " + r.exchange)
}

func (r *Relay) drain() {
	if !r.connected() {
		return
	}

	events, err := r.repository.ClaimEvents(relayBatch, relayLease)
	if err != nil {
		fmt.Println("claiming outbox events:", err)

		return
	}

	for _, event := range events {
		if err := r.publish(event); err != nil {
			fmt.Printf("publishing event %s: %v\n", event.EventID, err)

			r.fail(event, err)
			r.disconnect()

			return
		}

		if err := r.repository.MarkPublished(event.ID); err != nil {
			fmt.Println("marking event as published:", err)
		}
	}
}

func (r *Relay) connected() bool {
	if r.connection != nil && !r.connection.Closed() {
		return true
	}

	connection, err := Connect()
	if err != nil {
		return false
	}

	fmt.Println("Successfully connected to RabbitMQ")

	r.connection = connection

	return true
}

func (r *Relay) disconnect() {
	if r.connection == nil {
		return
	}

	r.connection.Close()
	r.connection = nil
}

func (r *Relay) publish(event model.OutboxEvent) error {
	ctx, cancel := context.WithTimeout(context.Background(), publishTimeout)
	defer cancel()

	confirmation, err := r.connection.channel.PublishWithDeferredConfirmWithContext(
		ctx,
		r.exchange,
		event.RoutingKey,
		true,
		false,
		amqp.Publishing{
			MessageId:     event.EventID,
			CorrelationId: event.CausationID,
			ContentType:   "application/json",
			DeliveryMode:  amqp.Persistent,
			Timestamp:     event.CreatedAt,
			Body:          event.Payload,
		},
	)
	if err != nil {
		return err
	}

	acknowledged, err := confirmation.WaitContext(ctx)
	if err != nil {
		return err
	}

	if !acknowledged {
		return errBrokerRefused
	}

	select {
	case returned := <-r.connection.returns:
		return fmt.Errorf("%w: %s", errUnroutable, returned.RoutingKey)
	default:
	}

	return nil
}

func (r *Relay) fail(event model.OutboxEvent, cause error) {
	if err := r.repository.MarkFailed(event.ID, cause.Error(), time.Now().Add(backoff(event.Attempts))); err != nil {
		fmt.Println("marking event as failed:", err)
	}
}

func backoff(attempts int) time.Duration {
	delay := backoffBase << attempts

	if delay > backoffCap || delay <= 0 {
		return backoffCap
	}

	return delay
}

File: finance/app/Messaging/Relay.php
<?php

namespace App\Messaging;

use App\Models\OutboxEvent;
use Closure;
use Illuminate\Support\Carbon;
use Illuminate\Support\Facades\Log;
use PhpAmqpLib\Message\AMQPMessage;
use Throwable;

class Relay
{
    private ?Connection $connection = null;

    public function __construct(
        private readonly Closure $log,
        private readonly Closure $sleep,
    ) {}

    public static function make(?Closure $log = null, ?Closure $sleep = null): self
    {
        return new self(
            $log ?? fn (string $line) => Log::info($line),
            $sleep ?? fn (int $seconds) => sleep($seconds),
        );
    }

    public function run(): void
    {
        $interval = (int) config('rabbitmq.relay.interval_seconds');

        $signals = Signals::trap();

        $this->report('outbox relay started, publishing to '.Topology::FINANCE_EXCHANGE);

        while (! $signals->stopping()) {
            $this->drain();

            if ($signals->stopping()) {
                break;
            }

            ($this->sleep)($interval);
        }

        $this->disconnect();
        $this->report('outbox relay stopped');
    }

    public function drain(): void
    {
        if (! $this->connected()) {
            return;
        }

        $events = OutboxEvent::claim(
            (int) config('rabbitmq.relay.batch'),
            (int) config('rabbitmq.relay.lease_seconds'),
        );

        foreach ($events as $event) {
            try {
                $this->publish($event);
            } catch (Throwable $exception) {
                $this->report("publishing event {$event->event_id}: {$exception->getMessage()}");
                $this->fail($event, $exception);
                $this->disconnect();

                return;
            }

            $event->markPublished();
        }
    }

    private function connected(): bool
    {
        if ($this->connection !== null && $this->connection->isOpen()) {
            return true;
        }

        try {
            $this->connection = Connection::open();
        } catch (Throwable $exception) {
            $this->report('connecting to RabbitMQ: '.$exception->getMessage());

            return false;
        }

        $this->report('successfully connected to RabbitMQ');

        return true;
    }

    private function disconnect(): void
    {
        $this->connection?->close();
        $this->connection = null;
    }

    private function publish(OutboxEvent $event): void
    {
        $properties = array_filter([
            'message_id' => $event->event_id,
            'correlation_id' => $event->causation_id,
            'content_type' => 'application/json',
            'delivery_mode' => AMQPMessage::DELIVERY_MODE_PERSISTENT,
            'timestamp' => $event->created_at->getTimestamp(),
        ], fn (mixed $value): bool => $value !== null);

        $this->connection?->publish(
            new AMQPMessage((string) json_encode($event->payload), $properties),
            Topology::FINANCE_EXCHANGE,
            $event->routing_key,
            (float) config('rabbitmq.relay.publish_timeout_seconds'),
        );
    }

    private function fail(OutboxEvent $event, Throwable $cause): void
    {
        $event->markFailed($cause->getMessage(), Carbon::now()->addSeconds($this->backoff($event->attempts)));
    }

    private function backoff(int $attempts): int
    {
        $base = (int) config('rabbitmq.relay.backoff_base_seconds');
        $cap = (int) config('rabbitmq.relay.backoff_cap_seconds');

        $delay = $base * (2 ** $attempts);

        return $delay > $cap || $delay <= 0 ? $cap : (int) $delay;
    }

    private function report(string $line): void
    {
        ($this->log)($line);
    }
}
